src/Sentry/MetricAggregator.cs

```diff
@@ -216,6 +214,44 @@ static string EscapeString(string input, params char[] charsToEscape)
         }
     }
 
+    private ConcurrentDictionary<string, Metric> GetOrAddTimeBucket(long bucketKey)
+    {
+        _bucketsLock.EnterUpgradeableReadLock();
+        try
+        {
+            if (Buckets.TryGetValue(bucketKey, out var existingBucket))
+            {
+                return existingBucket;
+            }
+            else
+            {
+                _bucketsLock.EnterWriteLock();
+                try
+                {
+                    // Check again in case another thread added the bucket while we were waiting for the write lock
+                    if (Buckets.TryGetValue(bucketKey, out existingBucket))
+                    {
+                        return existingBucket;
+                    }
+                    else
+                    {
+                        var timeBucket = new ConcurrentDictionary<string, Metric>();
+                        Buckets[bucketKey] = timeBucket;
+                        return timeBucket;
+                    }
+                }
+                finally
+                {
+                    _bucketsLock.ExitWriteLock();
+                }
+            }
```

Comment:
nit: maybe we could decrease the depth here a bit, e.g. 
```suggestion

            _bucketsLock.EnterWriteLock();
            try
            {
                // Check again in case another thread added the bucket while we were waiting for the write lock
                if (Buckets.TryGetValue(bucketKey, out existingBucket))
                {
                    return existingBucket;
                }
                
                var timeBucket = new ConcurrentDictionary<string, Metric>();
                Buckets[bucketKey] = timeBucket;
                return timeBucket;
            }
            finally
            {
                _bucketsLock.ExitWriteLock();
            }
```